Add initial machine learning pipeline

redshiftzero opened the pull request:
This PR adds an initial machine learning pipeline that takes the features in the database, trains a series of binary classifiers, evaluates how well each classifier performs, and then saves a bunch of relevant performance metrics in the database as well as pickling the trained model objects (for use in future scoring). The work in this PR corresponds to the latter half of this diagram from the `features` schema on: 
<img width="958" alt="pipeline" src="https://cloud.githubusercontent.com/assets/7832803/19282126/cec03c02-8fa1-11e6-9065-b46b3f321161.png">

A more complete description of our pipeline is described in `docs/pipeline.md` and a (very) brief description of how specialized classifiers might be integrated is stored in `CONTRIB.md`.

redshiftzero committed:
Specify evaluation type in YAML file

redshiftzero committed:
Refactor engine creation and add initial DatasetLoader class

redshiftzero committed:
Add generic function for getting feature importances of trained model

redshiftzero committed:
Add function for plotting feature importances

redshiftzero committed:
Add function for ROC plot

redshiftzero committed:
Add machine learning classifier and CV code, add Experiment() classifier wrapper

redshiftzero committed:
Add function to load closed world dataset

redshiftzero committed:
Add scikit-learn classifiers and hyperparameters

redshiftzero committed:
Add database setup for model storage

redshiftzero committed:
Add model storage to database.py

redshiftzero committed:
Add observed_fraction to open world validation

redshiftzero committed:
Add requirements for machine learning codes

redshiftzero committed:
Add little guide describing how to integrate a new classifier into the analysis pipeline

redshiftzero committed:
Add feature scaling option to attack setup

redshiftzero committed:
Add figures to documentation directory

redshiftzero committed:
Add writeup/readme on how pipeline works

redshiftzero committed:
Add tests for custom evaluation code

redshiftzero committed:
Add a ton of evaluation metrics to the models table

psivesely commented on docs/Pipeline.md:
````diff
@@ -0,0 +1,85 @@
+
+Run this step with:
+
+```python features.py```
````
This fails because the default Python in Xenial is 2.7. `./features.py` will read the shebang and choose the version appropriately. Be sure to fix the ones below too.

redshiftzero replied:
Word - changed. Also changed attack.py below

psivesely commented on docs/Pipeline.md:
```diff
@@ -0,0 +1,85 @@
+
+* `num_kfolds`: value of k for k-fold cross-validation 
+
+* `feature_scaling`: this option will take the features and rescale them to a zero mean and unit standard deviation. For some classifiers, e.g. primarily those based on decision trees, this should not improve performance, but for many classifiers, e.g. SVM, this is necessary. 
```
Proposal to help the ML newbies like me understand what's going on here:

> `feature_scaling`: setting [feature scaling](https://en.wikipedia.org/wiki/Feature_scaling) will take the features and compute their [Standard score](https://en.wikipedia.org/wiki/Standard_score), re-scaling them to zero.... See also [scikit-learn's documentation](http://scikit-learn.org/stable/modules/preprocessing.html).

redshiftzero replied:
Added these links!

psivesely commented on docs/Pipeline.md:
```diff
@@ -0,0 +1,85 @@
+* `auc`: [Area under the ROC curve](http://people.inf.elte.hu/kiss/12dwhdm/roc.pdf)
+* `tpr`: true positive rate [array for default sklearn thresholds]
+* `fpr`: false positive rate [array for default sklearn thresholds]
+* `precision_at_k` for `k=[0.01, 0.05, 0.1, 0.5, 1, 5, 10]`: "Fraction of SecureDrop users correctly identified in the top k percent"
```
Percent of what? I'll probably figure out later as I read on in the docs or source, but would be better if it was clarified here.

redshiftzero replied:
Percentage of the testing set, explicitly added

psivesely commented on fpsd/attack.py:
```diff
@@ -0,0 +1,79 @@
+
+    Args:
+        options [dict]: attack setup file
+    """
```
The last code I wrote (see https://github.com/freedomofpress/fingerprint-securedrop/blob/fecb7fa21811e6f40d66fb097880b02807c0375e/fpsd/utils.py#L28), I wrote with the documentation style described in #53. I'm not absolutely set on one particular style, but we should be consistent. If you want to make other suggestions, do so in #53 and we can discuss, but if you also like the python-gnupg style, then you should make the appropriate changes here.

redshiftzero replied:
I was using [Google style docstrings](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google), but don’t have strong feelings one way or the other so the docstring format you like there is fine w/me and from now on I will follow this format. However, most of the existing docstrings in this branch I wrote before that issue was filed so would rather not rewrite the Google style docstrings at this stage in the interests of time.

psivesely replied:
Sgtm.

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import pickle
+from scipy import interp
```
Don't see you using this dependency, and believe you meant to import the `interpolate` sub-package anyway.

redshiftzero replied:
Removed this.

psivesely commented on fpsd/database.py:
```diff
@@ -26,24 +28,139 @@ def safe_session(engine):
         session.close()
 
 
+def get_engine():
```
I implemented this functionality in a `Database` base class. You can get rid of this function, and replace line 52 below with `super().__init__(**kwargs)` .

redshiftzero replied:
Done

psivesely commented on fpsd/database.py:
```diff
@@ -26,24 +28,139 @@ def safe_session(engine):
+           df [pandas DataFrame]: dataset
+        """
+
+        labelled_query = ('select t1.*, t3.is_sd from features.frontpage_features t1 '
```
In Murrica we spell this labeled, but I just learned that spelling is normal for most other English speaking countries.

redshiftzero replied:
when in Rome I guess  ¯_(ツ)_/¯. Fixed

psivesely commented on fpsd/attack.py:
```diff
@@ -0,0 +1,79 @@
+
+    db = database.DatasetLoader()
+
+    if options["world"]["type"] == "closed":
```
The `load_closed_world` and `load_open_world` methods could easily be condensed into a single method by using format with query string. Then you could replace this block with the single line `df.db_closed_world()`.

redshiftzero replied:
OK, condensed this into one function

psivesely commented on fpsd/attack.py:
```diff
@@ -0,0 +1,79 @@
+    y = df['is_sd'].astype(int).values
+
+    experiments = generate_experiments(options)
+    for experiment in experiments:
```
I would condense this into one line `... in generate_experiments(options)`.

redshiftzero replied:
Fixed

psivesely commented on fpsd/attack.py:
```diff
@@ -0,0 +1,79 @@
+
+        parameter_names = sorted(model_hyperparameters)
+        parameter_values = [model_hyperparameters[p] for p in parameter_names]
+        all_params = product(*parameter_values)
```
At first I though this was the multiplicative product, a note specifying it's the Cartesian product would help readability.

redshiftzero replied:
Done

psivesely commented on fpsd/attack.py:
```diff
@@ -0,0 +1,79 @@
+
+if __name__=='__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-a", "--setup-yml", dest="attack_setup", type=str, default="attack.yaml", 
```
This argument naming scheme is a little funky. Since there's no conflicts to worry about, why not `-c, --config`? The help string should make it clear.

redshiftzero replied:
OK changed to config

psivesely commented on fpsd/requirements/ml-requirements.txt:
```diff
@@ -0,0 +1,14 @@
+seaborn==0.7.1
+SQLAlchemy==1.0.14
+tqdm==4.8.4
+
```
Can you put the top-level imports only in the `fpsd/requirements/requirements.in` file?

redshiftzero replied:
Done

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+class Experiment(object):
+    def __init__(self, model_timestamp, world, model_type, 
+                 hyperparameters,
+                 feature_scaling=True, n_cores=2, k=10):
```
I think setting `self.n_cores = multiprocessing.cpu_count()` in `__init__` is probably better than assuming there's 2.

redshiftzero replied:
Good call, added this

psivesely commented on fpsd/attack.yaml:
```diff
@@ -0,0 +1,90 @@
+  KNeighborsClassifier:
+    n_neighbors: [1, 3, 5, 10, 25, 50, 100]
+    weights: ['uniform', 'distance']
+    algorithm: ['auto', 'kd_tree']
```
What editor are you using that does this weird no newline at end of line thing?

redshiftzero replied:
I used Sublime Text to write this ¯_(ツ)_/¯.  Added newline

psivesely commented on fpsd/attack.yaml:
```diff
@@ -0,0 +1,90 @@
+
+world: 
+  type: 'open'  # 'closed' or 'open'
+  observed_fraction: 0.20  # fraction of the open world that can be measured by the adversary 
```
It might be fun to make this a hyperparameter to play with as well. Would help in threat modeling. Though it is computationally costly. So it could just be set to `[0.20]` initially, and then if we ever want to investigate later, we can add some other values.

redshiftzero replied:
Yep we can make this take a list at a later point, updated #64 issue to keep this suggestion

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+    return df.fillna(0)
+
+
+class Experiment(object):
```
You can simplify this to `class Experiment:`.

redshiftzero replied:
Fixed

psivesely commented on fpsd/requirements/ml-requirements.txt:
```diff
@@ -0,0 +1,14 @@
+ipython==5.1.0
```
This should not be here. It's installed by default except in Travis by an Ansible task.

redshiftzero replied:
Removed

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+            cv = cross_validation.StratifiedKFold(y_val, n_folds=self.k,
+                                                  shuffle=True)
+        elif self.world_type == "open":
+            pass  # TODO
```
Do it! (#64)

redshiftzero replied:
It was recommended that I make smaller PRs ^_^ this is why I created issue #64 and then submitted this PR before implementing this option. After this is merged, I will implement this in a separate PR. Trying to do the smaller PR thing to make incremental changes that are easier to review, but this means there are some TODO placeholders in the code that we will have to accept for now

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+
+        if self.model_type == "RandomForest":
+            return ensemble.RandomForestClassifier(
+                n_estimators=self.hyperparameters['n_estimators'],
```
Why not just unpack the `self.hyperparameters` wherever you can below (i.e., when the key names for the values you wish to pass line up with the corresponding arguments)?

redshiftzero replied:
Here is my concern: I think this is already difficult to understand for people not very familiar with scikit-learn classifiers and their hyperparameters. I'm worried that by using unpacking here this will become even more difficult to understand - it reduces the number of lines of code, but also negatively impacts readability. Thoughts?

psivesely replied:
Your call/ @conorsch merged this without me already, so it's in.

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+        if self.world_type == "closed":
+            # Why we use stratified k-fold here:
+            # http://stats.stackexchange.com/questions/49540/understanding-stratified-cross-validation
+            cv = cross_validation.StratifiedKFold(y_val, n_folds=self.k,
```
Did you consider using `sklearn.cross_validation.ShuffleSplit` here to address #61? Thoughts either way?

redshiftzero replied:
Not sure how this would address #61, but it’s not the best cross validation method to use here as it we want to make sure we are preserving the class balance as well as the fact that we’d have to specify the size of the training set, which `cross_validation.StratifiedKFold` does automatically by our specifying `k`

psivesely replied:
Precisely because it allows us to to balance representation as we like, which is something roughly suggested in #61.

>ShuffleSplit is thus a good alternative to KFold cross validation that allows a finer control on the number of iterations and the proportion of samples on each side of the train / test split.

--http://scikit-learn.org/stable/modules/cross_validation.html#random-permutations-cross-validation-a-k-a-shuffle-split

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+
+    def pickle_results(self, pkl_file, to_save):
+        with open(pkl_file, 'wb') as f:
+            pickle.dump(to_save, f, protocol=pickle.HIGHEST_PROTOCOL)
```
We should talk about the suggestions in http://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations. I'm personally in favor of pinning our version of `sklearn` in the `.in` file, and saving all the metadata they recommend along with the pickle modeled. If a new version of `sklearn` comes out with some fantastic features we want to use, it won't be the end of the world to retrain all our old models.

redshiftzero replied:
Lots of good comments in the link you share here. Some of the metadata in there is being saved (e.g. cross-validation scores) in the database, however I have now pinned version 0.17.0 of scikit-learn in requirements.in.

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+        trained_model = modelobj.fit(train_x, train_y)
+        return trained_model
+
+    def score(self, test_x, trained_model):
```
Can you keep the var as `x_test` in this fn. for consistency?

redshiftzero replied:
Fixed

psivesely commented on docs/Pipeline.md:
```diff
@@ -0,0 +1,85 @@
+
+## Model Selection
+
+This is currently done manually by selecting the top `auc` model in the database. 
```
I think in WTF-PAD they correctly point out that precision and recall are more important measures than TPR and FPR. They considered F<sub>1</sub> to be the most important metric, but I personally believe that F<sub>0.5</sub> is closer to the most informative single metric we can look at given our specific problem. There is a function `sklearn.metrics.fbeta_score` that will compute F<sub>B</sub>.

redshiftzero replied:
It’s true that precision, recall, and F1 score are better metrics, but for this first pass we are using AUC since it’s independent of class balance in the testing set (see Figure 5 in [this paper](http://people.inf.elte.hu/kiss/12dwhdm/roc.pdf)). I filed #62 to implement precision, recall, F1 storage. Not following why F_{0.5} would be a better metric?

psivesely replied:
Let's discuss in-person.

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+
+    for threshold in THRESHOLDS:
+        precision, recall, f1 = precision_recall_at_x_proportion(truth,
+            predicted, x_proportion=threshold/100.)
```
Can you drop the weird period? I know it has no effect, but...

redshiftzero replied:
Done

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+    eval_metrics.update({"auc": track_metric})
+
+    for metric in ("fpr", "tpr"):
+        # the fpr, tpr output from scikit-learn may not have the same
```
I don't understand this comment or what you're doing here. Can you please explain?

redshiftzero replied:
The TPR and FPR output from each individual fold need to be averaged (not _quite_ that simple since they are arrays that need to be interpolated onto a common threshold grid and then averaged) and saved. This is a TODO (added inline TODO) and is issue #65

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+    return eval_metrics
+
+
+def get_average_metrics(metrics_list):
```
You could condense this function significantly by using the one-liner

```
statistics.mean(map(lambda x: x[metric], metrics_list))
```

to compute the mean of a given metric over the folds.

redshiftzero replied:
Hm maybe I’m being dense right now but this doesn’t seem to work for me due to the scope here (I get `’metric’ is not defined`)?

psivesely replied:
Might need to change a variable name... would much of the last suite starting at line 63 I believe.

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+    eval_metrics["tpr"] = tpr
+    eval_metrics["thresholds"] = thresholds
+    eval_metrics["auc"] = metrics.auc(fpr, tpr)
+    eval_metrics["precision"] = {}
```
Why are you creating blank entries for these?

redshiftzero replied:
Placeholders to be implemented in #62, added inline TODO

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+    return fig
+
+
+def precision_recall_at_x_proportion(test_labels, test_predictions, x_proportion=0.01,
```
No idea what this function is doing or why. Needs at least a docstring, maybe a comment in `get_metrics` as well. Also, can you move this function right below `get_metrics` since they are related and the two functions above this one are not related to `get_metrics`?

redshiftzero replied:
Added docstring, moved up in the file

psivesely commented on fpsd/classify.py:
```diff
@@ -0,0 +1,277 @@
+        self.feature_scaling = feature_scaling
+        self.db = database.ModelStorage()
+
+    def get_dict(self):
```
There's a method that does this, `self.__dict__`. I would just rename the object attributes to match `sklearn`'s. More clear that way anyway.

redshiftzero replied:
Done in 5a474343

psivesely commented on fpsd/database.py:
```diff
@@ -26,24 +28,139 @@ def safe_session(engine):
+    return engine
+
+
+class DatasetLoader(object):
```
`class DatasetLoader(Database)`--classes below too.

redshiftzero replied:
Done

psivesely commented on fpsd/database.py:
```diff
@@ -26,24 +28,139 @@ def safe_session(engine):
+                    options["world_type"], options["model_type"],
+                    options["base_rate"], json.dumps(options["hyperparameters"]),
+                    self.metric_formatter(eval_metrics)))
+        with safe_session(self.engine) as session:
```
`with self.safe_session() as session:`--and below too.

redshiftzero replied:
Done

psivesely commented on fpsd/evaluation.py:
```diff
@@ -0,0 +1,193 @@
+def get_feature_importances(model):
+
+    try:
+        return model.feature_importances_
```
This attribute is cool. If we can identify features that are consistently found not to be useful for our top classifiers, we could use this information to help us implement the feature selection stage described in #63.

redshiftzero replied:
:+1:

psivesely reviewed (dismissed):
After changes and a rebase are made, I'll take another look. I'll do my part to help us iterate faster than happened with the features branch.

redshiftzero committed:
Feedback from code review

redshiftzero committed:
Move requirements

redshiftzero committed:
Add Ansible configuration of models schema

redshiftzero committed:
Add new pip requirements from pip-compile.sh

redshiftzero committed:
Point database base class to use the PGPASSFILE

Database needs a password, so I'm not sure how this was working before?
Either way, it's there now, and reads from the environmental variable
$PGPASSFILE.

redshiftzero committed:
Label column should always be is_sd

redshiftzero committed:
Database classes should use test keyword

redshiftzero committed:
Add test keyword in database class

The existing codes were always using the test database, which is not want we want.
The production or test database can be selected using this new test keyword.

redshiftzero committed:
Replace get_dict() with __dict__

redshiftzero commented:
OK comments addressed, ansible-ification of the creation of the models schema and tables is done, Travis builds are passing, and it's rebased on current master. Should be good to go 🌞

conorsch commented:
What a review process this has been! Thanks for your patience here, @redshiftzero. Given the frequent back-and-forth here, I'm inclined to merge, and we can bite off smaller hunks to discuss in discrete issues going forward.

redshiftzero commented:
👍  sounds good - any other outstanding issues we can make issues for and address in smaller PRs

conorsch reviewed (approved):
Changes requested during review have been implemented. :shipit:
